nixos/acme: Various fixes

m1cr0man opened the pull request:
###### Description of changes

Closes #191794 - Adds an AmbientCapabilities line to permit binding on port 80

Closes #190493 - Adds a check for an actual account key file rather than any file/folder in accounts

~WIP #180980 - Will fix the web server config for nginx.~ No response will fix in another PR.

CC @NixOS/acme 

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

m1cr0man committed:
nixos/acme: Fix cert renewal with built in webserver

Fixes #191794

Lego threw a permission denied error binding to port 80.
AmbientCapabilities with CAP_NET_BIND_SERVICE was required.
Also added a test for this.

m1cr0man committed:
nixos/acme: Make account creds check more robust

Fixes #190493

Check if an actual key file exists. This does not
completely cover the work accountHash does to ensure
that a new account is registered when account
related options are changed.

kevincox commented on nixos/tests/common/acme/client/default.nix:
```diff
@@ -9,6 +9,11 @@ in {
+      # We are not using LE servers in testing so this is not
+      # going to impact their load.
+      # See https://github.com/go-acme/lego/issues/1656
+      extraLegoRenewFlags = ["-no-random-sleep"];
```
Since we set `RandomizedDelaySec=24h` It might be best to pass this by default. This way we aren't starting the CLI and holding resources open for a delay. Plus if people manually remove `RandomizedDelaySec` they probably don't want the delay anyways.

Side note: We should probably also set `FixedRandomDelay=true` so that we are renewing on a consistent interval rather than just averaging the right interval. For daily renewals it probably doesn't matter much but if you are doing weekly and get unlucky you may be waiting nearly 2 weeks for the next check.

Garmelon replied:
But `RandomizedDelaySec` is only 24 hours and doesn't delay the next time the timer is triggered, so the worst case would be `renewInterval` + 24 hours in-between checks (e.g. 8 days for `weekly`)? `FixedRandomDelay` shouldn't change the (average) time between triggers because they're set via `OnCalendar`

kevincox replied:
Ah right. It can add at most 24h between triggers (last one rolls a 0h and the next one rolls a 23h59min). But I think it still makes sense to do `FixedRandomDelay` so that the time between is fairly consistent.

m1cr0man replied:
Yeah, agreed on adding no-random-sleep to the default options. To avoid triggering renewals on all upgrading systems, I am hard-coding it into the module's `renewFlags`. A random sleep during the execution of lego will needlessly delay the runtime of the service thus also delaying startup times for dependent systemd services, which is definitely unwanted in all cases.

I'll add FixedRandomDelay for the sake of it too.

m1cr0man committed:
nixos/acme: Disable lego renew sleeping

Lego has a built-in mechanism for sleeping for a random amount
of time before renewing a certificate. In our environment this
is not only unnecessary (as our systemd timer takes care of it)
but also unwanted since it slows down the execution of the
systemd service encompassing it, thus also slowing down the
start up of any services its depending on.

Also added FixedRandomDelay to the timer for more predictability.

m1cr0man commented:
Implemented the changes discussed in review comments. I am marking this as ready for review now since I don't know what the best course of action is to resolve #180980 for now, and will fix it at a later date.

winterqt commented on nixos/modules/security/acme/default.nix:
```diff
@@ -325,6 +325,7 @@ let
         '');
       } // optionalAttrs (data.listenHTTP != null && toInt (elemAt (splitString ":" data.listenHTTP) 1) < 1024) {
         CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" ];
+        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
```
It's my understanding that `CapabilityBoundingSet` and `AmbientCapabilities` are exact opposites of each other -- that is, `CapabilityBoundingSet` is a list of capabilities that shouldn't ever be granted to the process, while `AmbientCapabilities` is a list of capabilities to grant to the service (see [this SO answer](https://unix.stackexchange.com/a/581337)).

This being said, we should remove the `CapabilityBoundingSet` definition here, as it only serves to confuse.

(Do you know why it was even set in the first place? It's the exact opposite of what we want in this condition, it seems. Guessing it was just a mistake and nobody actually used it until now, or something?)

m1cr0man replied:
They aren't exact opposites, no. CapabilityBoundingSet controls what caps a program is allowed to utilise even when running as root. It is a whitelist of capabilities, and defaults to allowing all if unset. Ambient capabilities grants caps that are otherwise root reserved to non root executions. Thus both still have their place.

It was likely added as part of the service hardening that was conducted at some point. I don't have a link to the commit or PR handy.

winterqt replied:
Thanks for the explanation, and sorry for the misunderstanding.

aanderse replied:
To be fair the wording in that SO post is somewhat ambiguous.

Garmelon replied:
AmbientCapabilities was added because of #191794

winterqt replied:
> AmbientCapabilities was added because of #191794

I'm talking about the other option, which was added in place of AmbientCapabilities in 65f1b8c6ae2f2cf6a13d77b98b42eba31eef0424

@m1cr0man Shouldn't the test in that PR have caught this earlier? 🤔

winterqt commented:
@ofborg test acme

winterqt commented on nixos/tests/common/acme/client/default.nix:
```diff
@@ -9,6 +9,11 @@ in {
+      # We are not using LE servers in testing so this is not
+      # going to impact their load.
+      # See https://github.com/go-acme/lego/issues/1656
+      extraLegoRenewFlags = ["-no-random-sleep"];
```
Considering this is removed in the subsequent commit, would it make sense to just not add it in this commit?

m1cr0man replied:
Hm, I'm normally pretty adverse to editing commits because it alerts all the related issues again when you force push. Also I think these commits show a good evolution of the feature?

winterqt commented:
re: "This does not completely cover the work accountHash does to ensure that a new account is registered when account related options are changed."

What is this referencing, the new mechanism?

m1cr0man commented:
> re: "This does not completely cover the work accountHash does to ensure that a new account is registered when account related options are changed."
> 
> What is this referencing, the new mechanism?

So accountHash works to ensure that we are putting accounts for particular configurations in their own folder. The check for an existing cert file is solving a different issue, that's what I'm implying. Does that make sense?

winterqt commented:
Yeah, good enough. :)

winterqt reviewed (approved):
I'm confused as to why the port 80 issue wasn't caught earlier, but I won't make this wait on an answer for that. Thanks as always, @m1cr0man!

m1cr0man commented:
> I'm confused as to why the port 80 issue wasn't caught earlier, but I won't make this wait on an answer for that. Thanks as always, @m1cr0man!

We didn't actually have an automated test for that configuration 😅 I added one in this PR though. Thanks for merging!
